sound/soc/sof/ipc4-topology.c

```diff
@@ -1661,29 +1623,27 @@ static int sof_ipc4_prepare_process_module(struct snd_sof_widget *swidget,
 	memcpy(cfg, &process->base_config, sizeof(struct sof_ipc4_base_module_cfg));
 	cfg += sizeof(struct sof_ipc4_base_module_cfg);
 
-	/* copy output format to configure data payload */
-	if (process->payload_with_output_fmt) {
-		memcpy(cfg, &process->output_format, sizeof(struct sof_ipc4_audio_format));
-		cfg += sizeof(struct sof_ipc4_audio_format);
-	}
+	fw_module = swidget->module_info;
 
 	for (i = 0; i < widget->num_kcontrols; i++) {
-		kc = &widget->kcontrol_news[i];
+		const struct snd_kcontrol_new *kc = &widget->kcontrol_news[i];
+		struct sof_ipc4_control_data *control_data;
+		struct snd_sof_control *scontrol;
+		struct soc_bytes_ext *sbe;
 
 		/* payload uses byte kcontrol */
 		if (widget->dobj.widget.kcontrol_type[i] != SND_SOC_TPLG_TYPE_BYTES)
 			continue;
+
 		sbe = (struct soc_bytes_ext *)kc->private_value;
 		scontrol = sbe->dobj.private;
```

Comment:
We are looping the asoc widgets' kcontrol list and doing dereference directly, Pierre questions if it is safe before:https://github.com/thesofproject/linux/pull/4121#discussion_r1068580516, isn't it still a problem now?

Reply:
If the kcontrol type is SND_SOC_TPLG_TYPE_BYTES we know exactly what is behind the private_value, but I will check if this can be made cleaner.

Reply:
This part is now gone.